internal/handler/authenticate.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handler
+
+import (
+	"context"
+	"errors"
+
+	"github.com/FerretDB/FerretDB/internal/clientconn/conninfo"
+	"github.com/FerretDB/FerretDB/internal/handler/common"
+	"github.com/FerretDB/FerretDB/internal/handler/handlererrors"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+	"github.com/FerretDB/FerretDB/internal/util/password"
+)
+
+// authenticate validates users with stored credentials in admin.systems.user.
+// If EnableNewAuth is false or bypass backend auth is set false, it succeeds
+// authentication and let backend handle it.
+//
+// When admin.systems.user contains no user, the authentication is delegated to
+// the backend. This may change once local exception is implemented.
+// TODO https://github.com/FerretDB/FerretDB/issues/4100
+func (h *Handler) authenticate(ctx context.Context) error {
+	if !h.EnableNewAuth {
+		return nil
+	}
+
+	if !conninfo.Get(ctx).BypassBackendAuth() {
+		return nil
+	}
+
+	adminDB, err := h.b.Database("admin")
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	usersCol, err := adminDB.Collection("system.users")
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	username, pwd := conninfo.Get(ctx).Auth()
+
+	filter := must.NotFail(types.NewDocument("user", username))
+
+	qr, err := usersCol.Query(ctx, nil)
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	defer qr.Iter.Close()
+
+	var storedUser *types.Document
+
+	var hasUser bool
+
+	for {
+		var v *types.Document
+		_, v, err = qr.Iter.Next()
+
+		if errors.Is(err, iterator.ErrIteratorDone) {
+			break
+		}
+
+		if err != nil {
+			return lazyerrors.Error(err)
+		}
+
+		hasUser = true
```

Comment:
So, `hasUser` means that there is at least some user in the database, not necessarily the current user? And if there is at least one user in the database, we consider that new authentication is wanted, so we don't authenticate against the backend. Sounds reasonable!

Reply:
Correct. It's rather unpleasant work around.